File: LeetCode/0054.go
package leetcode

func spiralOrder(matrix [][]int) []int {
	if len(matrix) == 0 || len(matrix[0]) == 0 {
		return nil
	}
	t, b, left, right, tol := 0, len(matrix)-1, 0, len(matrix[0])-1, len(matrix)*len(matrix[0])
	ans := make([]int, tol)

	for num := 0; num < tol; {
		for j := left; j <= right; j++ {
			ans[num] = matrix[t][j]
			num++
		}
		t++
		for i := t; i <= b; i++ {
			ans[num] = matrix[i][right]
			num++
		}
		right--
		for j := right; j >= left && num < tol; j-- {
			ans[num] = matrix[b][j]
			num++
		}
		b--
		for i := b; i >= t && num < tol; i-- {
			ans[num] = matrix[i][left]
			num++
		}
		left++
	}
	return ans
}

File: LeetCode/0054.cpp
class Solution
{
public:
    vector<int> spiralOrder(vector<vector<int>> &matrix)
    {
        vector<int> res;
        int t = 0;
        int b = matrix.size() - 1;

        int l = 0;
        int r = matrix[0].size() - 1;
        int i = 0;
        int j = 0;
        int num = 0;
        int tar = matrix.size() * matrix[0].size();

        while (num < tar)
        {
            for (int i = l; i <= r; i++)
            {
                res.push_back(matrix[t][i]);
                num++;
            }
            t++;
            for (int i = t; i <= b; i++)
            {
                res.push_back(matrix[i][r]);
                num++;
            }
            r--;
            for (int i = r; i >= l && num < tar; i--)
            {
                res.push_back(matrix[b][i]);
                num++;
            }
            b--;
            for (int i = b; i >= t && num < tar; i--)
            {
                res.push_back(matrix[i][l]);
                num++;
            }
            l++;
        }
        return res;
    }
  
};